File: app/src/main/java/com/jakubwyc/mysecretvault/OpenVaultFragment.java
package com.jakubwyc.mysecretvault;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.Toast;

import com.jakubwyc.mysecretvault.presenter.OpenVaultPresenter;
import com.jakubwyc.mysecretvault.view.OpenVaultView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

public class OpenVaultFragment extends Fragment implements OpenVaultView {

    private OpenVaultPresenter presenter;
    private OpenVaultFragmentListener listener;
    private Unbinder unbinder;

    @BindView(R.id.login_edit)
    EditText loginView;
    @BindView(R.id.password_edit)
    EditText passwordView;

    @Nullable
    @Override
    public View onCreateView(final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_open_vault, container, false);
        unbinder = ButterKnife.bind(this, view);

        return view;
    }

    @Override
    public void onAttach(final Context context) {
        super.onAttach(context);
        presenter = new OpenVaultPresenter();
        presenter.attachView(this);
        if (context instanceof OpenVaultFragmentListener) {
            listener = (OpenVaultFragmentListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OpenVaultFragmentListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        presenter.detachView();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @OnClick(R.id.create_vault_button)
    @Override
    public void showCreateVaultScreen() {
        if (listener != null) {
            listener.onFragmentChange(MainActivity.VaultScreen.CREATE);
        }
    }

    @OnClick(R.id.open_vault_button)
    public void tryToOpenVault() {
        presenter.verifyUserCredentials(loginView.getText().toString(), passwordView.getText().toString());
    }

    @Override
    public void showVaultScreen() {
        if (listener != null) {
            listener.onFragmentChange(MainActivity.VaultScreen.OPEN);
        }
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }

    public interface OpenVaultFragmentListener {
        void onFragmentChange(MainActivity.VaultScreen screen);
    }
}

File: app/src/main/kotlin/com/jakubwyc/mysecretvault/vault/open/OpenVaultFragment.kt
package com.jakubwyc.mysecretvault.vault.open

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.jakewharton.rxbinding.view.RxView
import com.jakubwyc.mysecretvault.R
import com.jakubwyc.mysecretvault.SystemContext
import com.jakubwyc.mysecretvault.main.MainActivity
import kotlinx.android.synthetic.main.fragment_open_vault.*
import rx.subscriptions.CompositeSubscription

class OpenVaultFragment : Fragment(), OpenVaultView {

    private lateinit var presenter: OpenVaultPresenter
    private lateinit var listener: OpenVaultFragmentListener
    private val subscriptions = CompositeSubscription()

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_open_vault, container, false)
        return view
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        subscriptions.add(RxView.clicks(createVaultButton).subscribe { showCreateVaultScreen() })
        subscriptions.add(RxView.clicks(openVaultButton).subscribe { tryToOpenVault() })
    }

    override fun onResume() {
        super.onResume()
        clearUserData()
    }

    override fun clearUserData() {
        loginEdit.text.clear()
        passwordEdit.text.clear()
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        presenter = OpenVaultPresenter(activity.applicationContext as SystemContext)
        presenter.attachView(this)
        listener = context as? OpenVaultFragmentListener
                ?: throw RuntimeException(context.toString() + " must implement OpenVaultFragmentListener")
    }

    override fun onDetach() {
        super.onDetach()
        presenter.detachView()
        subscriptions.clear()
    }

    override fun showCreateVaultScreen() {
        listener.onFragmentChange(MainActivity.VaultScreen.CREATE)
    }

    fun tryToOpenVault() {
        presenter.verifyUserCredentials(loginEdit.text.toString(), passwordEdit.text.toString())
    }

    override fun showVaultScreen() {
        listener.onFragmentChange(MainActivity.VaultScreen.VAULT)
    }

    override fun showMessage(message: String) {
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    interface OpenVaultFragmentListener {
        fun onFragmentChange(screen: MainActivity.VaultScreen)
    }
}
